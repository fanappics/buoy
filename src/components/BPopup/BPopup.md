<docs>
BPopup on a Button

```js
let popupVisible = false

<div>
  <b-button @click="popupVisible = true">
    Click me
    <b-popup
      :visible="popupVisible"
      message="You clicked this button."
      @close="popupVisible = false"
    />
  </b-button>
</div>
```

Play with size

```js
let popupVisible = false

<div>
  <b-button @click="popupVisible = true">
    Click me
    <b-popup
      :visible="popupVisible"
      message="A slim popup."
      width="200"
      height="44"
      @close="popupVisible = false"
    />
  </b-button>
</div>
```

Play with position

```js
let popupVisible = false

<div>
  <b-button @click="popupVisible = true">
    Click me
    <b-popup
      :visible="popupVisible"
      message="Above the button."
      offset-x="-50"
      offset-y="-135"
      @close="popupVisible = false"
    />
  </b-button>
</div>
```

BPopup Confirmation Example

```js
let popupVisible = false
let confirmed = false

<div>
  <b-button @click="popupVisible = true">
    Delete all files
    <b-popup
      :visible="popupVisible"
      message="Are you sure this is what you want?"
      confirm
      @close="popupVisible = false"
      @confirm="confirmed = true"
    />
  </b-button>
  Confirmed: {{confirmed}}
</div>
```
</docs>
