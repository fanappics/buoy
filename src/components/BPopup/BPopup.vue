<template>
  <div v-if="visible" v-click-outside="close" class="b-popup">
    <div class="b-wrapper" :style="[size, offset]">
      <div :class="{ 'b-content': true }">
        <span>{{message}}</span>
      </div>
      <div v-if="confirm" class="b-footer">
        <button @click="cancelClickHandler">Cancel</button>
        <button @click="confirmClickHandler">Confirm</button>
      </div>
    </div>
  </div>
</template>

<script>
import clickOutside from '../../directives/click-outside'

export default {
  name: 'b-popup',
  props: {
    //Required props
    /**
    * The written message in the popup
    */
    message: {
      type: String,
      required: true
    },
    /**
    * When true the popup will be visible, not visible otherwise
    */
    visible: {
      type: Boolean,
      required: true
    },
    //Optional props
    /**
    * Add buttons to popup for confirmation purposes
    */
    confirm: Boolean,
    /**
    * X axis offset of popup in pixels
    */
    offsetX: {
      type: [String, Number],
      default: 0
    },
    /**
    * Y axis offset of popup in pixels
    */
    offsetY: {
      type: [String, Number],
      default: 0
    },
    /**
    * Height of popup in pixels
    */
    height: {
      type: [String, Number],
      default: 100
    },
    /**
    * Width of popup in pixels
    */
    width: {
      type: [String, Number],
      default: 150
    }
  },
  data () {
    return {
      size: {
        width: this.width,
        height: this.height
      },
      offset: {
        left: this.offsetX,
        top: this.offsetY
      }
    }
  },
  methods: {
    cancelClickHandler () {
      /**
      * Event fires when cancel is clicked on confirmation
      *
      * @event cancel
      */
      this.$emit('cancel')
      this.$nextTick(function () {
        this.close()
      })
    },
    close () {
      /**
      * Event fires when the popup finishes its dialog and recommends being closed
      *
      * @event close
      */
      this.$emit('close')
    },
    confirmClickHandler () {
      /**
      * Event fires when confirm is clicked on confirmation
      *
      * @event confirm
      */
      this.$emit('confirm')
      this.$nextTick(function () {
        this.close()
      })
    }
  },
  directives: clickOutside
}
</script>

<style lang="stylus" scoped>
  @import "../../styles/colors"

  .b-content
    padding .857rem

  .b-footer
    display flex
    justify-content center
    > button
      background-color transparent
      border none
      padding-top 0
    > button:focus
      color $cerulean
      outline none
    > button:hover
      color $cerulean
    > button:active
      border-style none

  .b-popup
    position relative
    z-index 50

  .b-wrapper
    background-color $white
    box-shadow 0 1px 2px 0
    cursor default
    display flex
    flex-direction column
    justify-content space-evenly
    position absolute

  div
    color $text
</style>
