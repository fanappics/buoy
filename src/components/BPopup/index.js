import BPopup from './BPopup.vue'

/* istanbul ignore next */
BPopup.install = function install (Vue) {
  Vue.component(BPopup.name, BPopup)
}

export default BPopup
