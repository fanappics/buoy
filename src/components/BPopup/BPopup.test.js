import { mount } from 'vue-test-utils'
import BPopup from './BPopup'

describe('BPopup', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = mount(BPopup, {
      propsData: {
        message: 'Sample popup message',
        visible: true
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('BPopup not visible', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = mount(BPopup, {
      propsData: {
        message: 'Sample popup message',
        visible: false
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('BPopup with confirmation buttons', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = mount(BPopup, {
      propsData: {
        message: 'Are you sure?',
        visible: true,
        confirm: true
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('BPopup with overriding size styling', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = mount(BPopup, {
      propsData: {
        message: 'Sample popup message',
        visible: true,
        width: '500',
        height: '900'
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('BPopup with offset styling', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = mount(BPopup, {
      propsData: {
        message: 'Sample popup message',
        visible: true,
        offsetX: '99',
        offsetY: '-99'
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
