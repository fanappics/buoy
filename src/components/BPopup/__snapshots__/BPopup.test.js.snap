// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BPopup mounted html should match snapshot 1`] = `
<div class="b-popup">
    <div class="b-wrapper" style="left: 0px; top: 0px;">
        <div class="b-content"><span>Sample popup message</span></div>
        <!---->
    </div>
</div>
`;

exports[`BPopup not visible mounted html should match snapshot 1`] = `undefined`;

exports[`BPopup with confirmation buttons mounted html should match snapshot 1`] = `
<div class="b-popup">
    <div class="b-wrapper" style="left: 0px; top: 0px;">
        <div class="b-content"><span>Are you sure?</span></div>
        <div class="b-footer">
            <button>Cancel</button>
            <button>Confirm</button>
        </div>
    </div>
</div>
`;

exports[`BPopup with offset styling mounted html should match snapshot 1`] = `
<div class="b-popup">
    <div class="b-wrapper">
        <div class="b-content"><span>Sample popup message</span></div>
        <!---->
    </div>
</div>
`;

exports[`BPopup with overriding size styling mounted html should match snapshot 1`] = `
<div class="b-popup">
    <div class="b-wrapper" style="left: 0px; top: 0px;">
        <div class="b-content"><span>Sample popup message</span></div>
        <!---->
    </div>
</div>
`;
